File: app/src/main/java/br/com/jonathanzanella/myexpenses/bill/ShowBillActivity.java
package br.com.jonathanzanella.myexpenses.bill;

import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.TextView;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.database.RepositoryImpl;
import br.com.jonathanzanella.myexpenses.expense.Expense;
import br.com.jonathanzanella.myexpenses.expense.ExpenseRepository;
import br.com.jonathanzanella.myexpenses.helpers.CurrencyHelper;
import br.com.jonathanzanella.myexpenses.views.BaseActivity;
import butterknife.BindView;

public class ShowBillActivity extends BaseActivity implements BillContract.View {
	public static final String KEY_BILL_UUID = "KeyBillUuid";

	@BindView(R.id.act_show_bill_name)
	TextView billName;
	@BindView(R.id.act_show_bill_amount)
	TextView billAmount;
	@BindView(R.id.act_show_bill_due_date)
	TextView billDueDate;
	@BindView(R.id.act_show_bill_init_date)
	TextView billInitDate;
	@BindView(R.id.act_show_bill_end_date)
	TextView billEndDate;

	private final BillPresenter presenter;

	public ShowBillActivity() {
		ExpenseRepository expenseRepository = new ExpenseRepository(new RepositoryImpl<Expense>(this));
		presenter = new BillPresenter(new BillRepository(new RepositoryImpl<Bill>(this), expenseRepository));
	}

	@Override
	protected void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_show_bill);
	}

	@Override
	protected void onPostCreate(Bundle savedInstanceState) {
		super.onPostCreate(savedInstanceState);
		presenter.onViewUpdated(false);
	}

	@Override
	protected void storeBundle(final Bundle extras) {
		super.storeBundle(extras);

		if(extras != null && extras.containsKey(KEY_BILL_UUID)) {
			new AsyncTask<Void, Void, Void>() {

				@Override
				protected Void doInBackground(Void... voids) {
					presenter.loadBill(extras.getString(KEY_BILL_UUID));
					return null;
				}

				@Override
				protected void onPostExecute(Void aVoid) {
					super.onPostExecute(aVoid);
					presenter.updateView();
				}
			}.execute();
		}
	}

	@Override
	protected void onSaveInstanceState(Bundle outState) {
		super.onSaveInstanceState(outState);
		outState.putString(KEY_BILL_UUID, presenter.getUuid());
	}

	@Override
	protected void onStart() {
		super.onStart();
		presenter.attachView(this);
	}

	@Override
	protected void onStop() {
		super.onStop();
		presenter.detachView();
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		getMenuInflater().inflate(R.menu.edit, menu);
		return true;
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
			case R.id.action_edit:
				Intent i = new Intent(this, EditBillActivity.class);
				i.putExtra(EditBillActivity.KEY_BILL_UUID, presenter.getUuid());
				startActivity(i);
				break;
		}
		return super.onOptionsItemSelected(item);
	}

	@Override
	public void showBill(Bill bill) {
		billName.setText(bill.getName());
		billAmount.setText(CurrencyHelper.format(bill.getAmount()));
		billDueDate.setText(String.valueOf(bill.getDueDate()));
		billInitDate.setText(Bill.SIMPLE_DATE_FORMAT.format(bill.getInitDate().toDate()));
		billEndDate.setText(Bill.SIMPLE_DATE_FORMAT.format(bill.getEndDate().toDate()));
	}
}

File: app/src/main/kotlin/br/com/jonathanzanella/myexpenses/bill/ShowBillActivity.kt
package br.com.jonathanzanella.myexpenses.bill

import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.widget.TextView
import br.com.jonathanzanella.myexpenses.App
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.helpers.toCurrencyFormatted
import br.com.jonathanzanella.myexpenses.transaction.Transaction
import br.com.jonathanzanella.myexpenses.views.anko.*
import org.jetbrains.anko.*
import javax.inject.Inject

class ShowBillActivity : AppCompatActivity(), BillContract.View {
    @Inject
    lateinit var presenter: BillPresenter
    override val context = this
    private val ui = ShowBillActivityUi()

    init {
        App.getAppComponent().inject(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ui.setContentView(this)

        storeBundle(savedInstanceState)
        storeBundle(intent.extras)

        ui.toolbar.setup(this)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        presenter.onViewUpdated(false)
    }

    fun storeBundle(extras: Bundle?) {
        if (extras?.containsKey(KEY_BILL_UUID) == true) {
            doAsync {
                presenter.loadBill(extras.getString(KEY_BILL_UUID))

                uiThread { presenter.updateView() }
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString(KEY_BILL_UUID, presenter.uuid)
    }

    override fun onStart() {
        super.onStart()
        presenter.attachView(this)
    }

    override fun onStop() {
        super.onStop()
        presenter.detachView()
    }

    override fun setTitle(string: String) {
        ui.toolbar.title = string
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.edit, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_edit -> {
                val i = Intent(this, EditBillActivity::class.java)
                i.putExtra(EditBillActivity.KEY_BILL_UUID, presenter.uuid)
                startActivity(i)
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun showBill(bill: Bill) {
        ui.apply {
            billName.text = bill.name
            billAmount.text = bill.amount.toCurrencyFormatted()
            billDueDate.text = bill.dueDate.toString()
            billInitDate.text = Transaction.SIMPLE_DATE_FORMAT.format(bill.initDate?.toDate())
            billEndDate.text = Transaction.SIMPLE_DATE_FORMAT.format(bill.endDate?.toDate())
        }
    }

    companion object {
        val KEY_BILL_UUID = "KeyBillUuid"
    }
}

class ShowBillActivityUi : AnkoComponent<ShowBillActivity> {
    lateinit var toolbar : TemplateToolbar
    lateinit var billName: TextView
    lateinit var billAmount: TextView
    lateinit var billDueDate: TextView
    lateinit var billInitDate: TextView
    lateinit var billEndDate: TextView

    override fun createView(ui: AnkoContext<ShowBillActivity>) = with(ui) {
        verticalLayout {
            toolbar = toolbarTemplate {}

            tableViewFrame {
                tableRow {
                    static { text = resources.getString(R.string.name) }
                    billName = staticWithData { id = R.id.act_show_bill_name }
                }
                tableRow {
                    static { text = resources.getString(R.string.amount) }
                    billAmount = staticWithData { id = R.id.act_show_bill_amount }
                }
                tableRow {
                    static { text = resources.getString(R.string.due_date) }
                    billDueDate = staticWithData { id = R.id.act_show_bill_due_date }
                }
                tableRow {
                    static { text = resources.getString(R.string.start) }
                    billInitDate = staticWithData { id = R.id.act_show_bill_init_date }
                }
                tableRow {
                    static { text = resources.getString(R.string.end) }
                    billEndDate = staticWithData { id = R.id.act_show_bill_end_date }
                }
            }
        }.applyRecursively(::applyTemplateViewStyles)
    }
}
